### PR TITLE
Refactor `Fix` and `Edit` API

### PR DESCRIPTION
**Pull request by madkinsz**
Part of https://github.com/charliermarsh/ruff/issues/4181
Closes https://github.com/charliermarsh/ruff/issues/4182

Thanks for the well specified issue! This implements many of the noted changes (see commit messages). There are a couple of deviations:

- Removes `Fix::empty`
- Updates `Message::fix` to be `Option<Fix>`
- Updates `MessageHeader::fix` to be `Option<Fix>`
- Does not remove `From<Edit> for Fix` (see [discussion](https://github.com/charliermarsh/ruff/pull/4198#issuecomment-1537473767))

### CODE REVIEW

File: crates/ruff/src/autofix/mod.rs [hunk 1]
```diff
@@ -38,10 +38,10 @@ fn apply_fixes<'a>(
...
     for (rule, fix) in diagnostics
         .filter_map(|diagnostic| {
-            if diagnostic.fix.is_empty() {
+            if diagnostic.fix.is_none() {
                 None
             } else {
-                Some((diagnostic.kind.rule(), &diagnostic.fix))
+                Some((diagnostic.kind.rule(), diagnostic.fix.as_ref().unwrap()))
             }
         })
```
Comment (madkinsz) [inline crates/ruff/src/autofix/mod.rs:46]: I'm guessing there's a more idiomatic way to handle this now that `diagnostic.fix` is an `Option`?

**Comment by madkinsz** [comment]
The removal of `From<Edit> for Fix` is pretty heavy. It looks like many rule functions passed to `try_set_fix` return `Edit` types. Is the intention to replace the return value of all of those functions with `Fix::unspecified`? e.g. https://github.com/charliermarsh/ruff/blob/cab65b25da8aff6fb9cb539b55345aae64a835f0/crates/ruff/src/rules/pyflakes/fixes.rs#L198

File: crates/ruff/src/autofix/mod.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff/src/autofix/mod.rs:46]: You can write `diagnostic.fix().map(|fix| (diagnostic.kind.rule(), fix))`

**Comment by MichaReiser** [comment]
> The removal of `From<Edit> for Fix` is pretty heavy. It looks like many rule functions passed to `try_set_fix` return `Edit` types. Is the intention to replace the return value of all of those functions with `Fix::unspecified`? e.g.
> 
> https://github.com/charliermarsh/ruff/blob/cab65b25da8aff6fb9cb539b55345aae64a835f0/crates/ruff/src/rules/pyflakes/fixes.rs#L198

Yeah, this is unfortunate but necessary because we'll need to explicitly pass the `Applicability` in the future.

**Review by MichaReiser — COMMENTED** [review]
Thank you for working on this refactor

File: crates/ruff/src/autofix/mod.rs (hunk 1 above)
Comment (madkinsz) [reply crates/ruff/src/autofix/mod.rs:46]: Wonderful thanks!

**Comment by madkinsz** [comment]
> Yeah, this is unfortunate but necessary because we'll need to explicitly pass the Applicability in the future.

Makes sense, just wanted to check before I really went down the rabbit hole on that one — I'll make those changes next and mark this as ready for review when it's done.

**Comment by madkinsz** [comment]
@MichaReiser I've encountered a bit of a problem and since the work is rather repetitive I'd like to check in again :)

In many cases, we can just replace `Result<Edit>` with `Result<Fix>` but there are generic functions like `delete_stmt` that return `Result<Edit>` and should probably continue to do so. Since they're used from many rules, I think it does not make sense for it to decide on the applicability of the proposed edits.
https://github.com/charliermarsh/ruff/blob/561f610b0543c28d46916c5760012c342aa8b08f/crates/ruff/src/autofix/actions.rs#LL169C1-L169C1

However there are some cases where this is used in `Diagnostic::try_set_fix`
https://github.com/charliermarsh/ruff/blob/e3bac78ea70667358a87ee090e2734bd9bb87dcb/crates/ruff/src/rules/flake8_pie/rules.rs#L143-L152

In this use, there's not a clear way to provide the specificity without requiring an `Applicability` to be passed to `try_set_fix` or introducing new factory methods like `Fix::try_unspecified` or `Diagnostics::try_set_unspecified_fix`. 

Suggestions?

Overall, this requires some careful editing as there are additional cases where a function is reused and probably should not decide the applicability level.

**Comment by MichaReiser** [comment]
> @MichaReiser I've encountered a bit of a problem and since the work is rather repetitive I'd like to check in again :)

I would have been surprised if my proposal just worked :laughing:. Thanks for doing this work so carefully. 

> In many cases, we can just replace `Result<Edit>` with `Result<Fix>` but there are generic functions like `delete_stmt` that return `Result<Edit>` and should probably continue to do so. Since they're used from many rules, I think it does not make sense for it to decide on the applicability of the proposed edits. [`561f610`/crates/ruff/src/autofix/actions.rs#LL169C1-L169C1](https://github.com/charliermarsh/ruff/blob/561f610b0543c28d46916c5760012c342aa8b08f/crates/ruff/src/autofix/actions.rs#LL169C1-L169C1)

I agree, that this should be decided by the caller rather than by `delete_stmt`. 

> 
> However there are some cases where this is used in `Diagnostic::try_set_fix`
> 
> https://github.com/charliermarsh/ruff/blob/e3bac78ea70667358a87ee090e2734bd9bb87dcb/crates/ruff/src/rules/flake8_pie/rules.rs#L143-L152
> 
> In this use, there's not a clear way to provide the specificity without requiring an `Applicability` to be passed to `try_set_fix` or introducing new factory methods like `Fix::try_unspecified` or `Diagnostics::try_set_unspecified_fix`.

In this case, I would create the `Fix` inside of the callback together, using the `try` operator. It's a bit more verbose but avoids the need for new factory methods. 

```rust
 diagnostic.try_set_fix(|| { 
	Ok(Fix::unspecified(
		delete_stmt( 
	         pass_stmt, 
	         None, 
	         &[], 
	         checker.locator, 
	         checker.indexer, 
	         checker.stylist, 
	     ) ?
	))
 }); 
```

> 
> Suggestions?
> 
> Overall, this requires some careful editing as there are additional cases where a function is reused and probably should not decide the applicability level. 

I think it's fine if we get some of them wrong because we have to re-visit all of them anyway when deciding whether we want to mark the fixes as `safe` or `MaybeIncorrect`. We can then decide whether the function should defer the determination of the fixe's safety to the caller or even change the signature to return an `Edit`.

The main motivation behind changing all of them to `unspecified` is that we can then use the IDE to find all places that still need to be changed.


> My hopes of replacing these with regular expressions have gone down the drain grin Maybe it'd be easier to leave `From<Edit> for Fix` in place for now and update each call site individually?

That's' also be a valid approach that I haven't thought of. Determining the "fixes to migrate" workflow could then be:

* comment out the `From` implementation
* fix all build errors of a single rule, 
* uncomment the `From` implementation except if there are no build errors, then remove the `From` implementation. 
* Submit PR
 
 I'll let you decide. You know the extent of the work best.

**Comment by madkinsz** [comment]
Of course you can just use a closure :) my lack of experience with Rust is showing. I think that's a reasonable approach if you think it's maintainable. I'll forge onward but if the changes feel hard to review I'll consider the "separate changes" workflow some more.

Thanks for the guidance again!

**Comment by madkinsz** [comment]
@MichaReiser I am leaning towards separate pull requests for individual rules. Otherwise this is going to take a while longer — it'll be harder to review and I'll need to keep rebasing as old patterns are used on `main`.

File: crates/ruff_cli/src/cache.rs [hunk 2]
```diff
@@ -233,7 +233,7 @@ pub fn get(
                 messages.push(Message {
                     kind: header.kind,
                     range: header.range,
-                    fix: header.fix,
+                    fix: Some(header.fix),
```
Comment (MichaReiser) [inline crates/ruff_cli/src/cache.rs:236]: `MessageHeader` is a helper struct to deserialize a `Message`. Its fields must match the fields on `Message` or the deserialization will fail if we have a `Message` without a header. 

That's why we should change the `fix` definition on `MessageHeader` to `Option<Fix>` to match `Message`

```rust 
#[derive(Deserialize)]
struct MessageHeader {
    kind: DiagnosticKind,
    range: TextRange,
    fix: Option<Fix>,
    file_id: usize,
    noqa_row: TextSize,
}
```

**Review by MichaReiser — COMMENTED** [review]
> @MichaReiser I am leaning towards separate pull requests for individual rules. Otherwise this is going to take a while longer. It'll be harder to review and I'll need to keep rebasing as old patterns are used on `main`. I also wouldn't be upset if you wanted to take a look — familiarity with the codebase would probably make doing it all at once significantly easier.

Sounds good to me. We can merge the PR as is and do a couple of follow-up PRs where we migrate a bunch of `.set_fix` calls to use `Fix::unspecified`. That also gives us the option for multiple collaborators to contribute to the refactor, so that you don't have to do all by yourself. 

Thank you

File: crates/ruff_cli/src/cache.rs (hunk 2 above)
Comment (madkinsz) [reply crates/ruff_cli/src/cache.rs:236]: Do you normally add tests for things like this to ensure it stays up to date?

**Comment by madkinsz** [comment]
Following merge, we can add a comment to #4181 and/or https://github.com/charliermarsh/ruff/issues/4184 covering the workflow for incremental removal of `From<Edit> for Fix`

File: crates/ruff_cli/src/cache.rs (hunk 2 above)
Comment (MichaReiser) [reply crates/ruff_cli/src/cache.rs]: Thx. We probably should but don't have to as part of this PR.

**Review by MichaReiser — APPROVED** [review]
Thank you, this is awesome.

**Pull request merged**